include/oneapi/dpl/experimental/kt/single_pass_scan.h

```diff
@@ -0,0 +1,434 @@
+    ::std::size_t __mem_bytes = __status_flags_bytes + __status_vals_full_offset_bytes +
+                                __status_vals_partial_offset_bytes + __mem_align_pad;
+
+    ::std::byte* __device_mem = reinterpret_cast<::std::byte*>(sycl::malloc_device(__mem_bytes, __queue));
+    assert(__device_mem);
+
+    _FlagStorageType* __status_flags = reinterpret_cast<_FlagStorageType*>(__device_mem);
+    ::std::size_t __remainder = __mem_bytes - __status_flags_bytes;
+    void* __vals_base_ptr = reinterpret_cast<void*>(__device_mem + __status_flags_bytes);
+    void* __vals_aligned_ptr = ::std::align(::std::alignment_of_v<_Type>, __status_vals_full_offset_bytes,
+                                            __vals_base_ptr, __remainder);
+    _Type* __status_vals_full = reinterpret_cast<_Type*>(__vals_aligned_ptr);
+    _Type* __status_vals_partial = reinterpret_cast<_Type*>(__status_vals_full + __status_vals_full_offset_bytes);
+
+    auto __fill_event = __lookback_init_submitter<_FlagType, _Type, _BinaryOp, _LookbackInitKernel>{}(
+        __queue, __status_flags, __status_vals_partial, __status_flags_size, __status_flag_padding);
+
+    ::std::size_t __current_num_wgs = oneapi::dpl::__internal::__dpl_ceiling_div(__n, __elems_in_tile);
+    ::std::size_t __current_num_items = __current_num_wgs * __workgroup_size;
+
+    auto __prev_event =
+        __lookback_submitter<__data_per_workitem, __workgroup_size, _Type, _FlagType, _LookbackKernel>{}(
+            __queue, __fill_event, __in_rng, __out_rng, __binary_op, __n, __status_flags, __status_flags_size,
+            __status_vals_full, __status_vals_partial, __current_num_items);
+
+    // TODO: Currently, the following portion of code makes this entire function synchronous.
+    // Ideally, we should be able to use the asynchronous free below, but we have found that doing
+    // so introduces a large unexplainable slowdown. Once this slowdown has been identified and corrected,
+    // we should replace this code with the asynchronous version below.
+    if (0)
+    {
+        return __queue.submit([=](sycl::handler& __hdl) {
+            __hdl.depends_on(__prev_event);
+            __hdl.host_task([=]() { sycl::free(__device_mem, __queue); });
+        });
+    }
+    else
+    {
+        __prev_event.wait();
+        sycl::free(__device_mem, __queue);
+        return __prev_event;
+    }
+}
+
+} // namespace __impl
+
+template <typename _InRng, typename _OutRng, typename _BinaryOp, typename _KernelParam>
+sycl::event
+inclusive_scan(sycl::queue __queue, _InRng&& __in_rng, _OutRng&& __out_rng, _BinaryOp __binary_op,
+               _KernelParam __param = {})
+{
+    auto __in_view = oneapi::dpl::__ranges::views::all(::std::forward<_InRng>(__in_rng));
+    auto __out_view = oneapi::dpl::__ranges::views::all(::std::forward<_OutRng>(__out_rng));
+
+    return __impl::__single_pass_scan<true>(__queue, ::std::move(__in_view), ::std::move(__out_view), __binary_op,
```

Comment:
I think we should use `std::forward` here instead of `std::move`

Reply:
Both `__in_view` and `__out_view` are created in this function directly above this line (they are not the ranges passed in). Since the function ends here, it doesn't hurt to explicitly move these views as they go out of scope after this line.